### PR TITLE
VSF2-159 Generate bulk data

### PR DESCRIPTION
**Pull request by klloz**

### CODE REVIEW

File: spec/DataGenerator/Generator/Entity/ProductGeneratorSpec.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    ): void {
+        $this->beConstructedWith($productFactory, $productVariantFactory, $channelPricingFactory);
+    }
+
+    public function it_is_initializable(): void
+    {
+        $this->shouldHaveType(ProductGenerator::class);
+    }
+
+    public function it_generates(
```
Comment (senghe) [inline spec/DataGenerator/Generator/Entity/ProductGeneratorSpec.php:41]: Where is any mock? This question applies for all entity generator specs.

File: spec/DataGenerator/Generator/Entity/TaxonGeneratorSpec.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+use BitBag\SyliusVueStorefront2Plugin\DataGenerator\Doctrine\Repository\TaxonRepositoryInterface;
+use BitBag\SyliusVueStorefront2Plugin\DataGenerator\Exception\InvalidContextException;
+use BitBag\SyliusVueStorefront2Plugin\DataGenerator\Factory\Entity\TaxonFactoryInterface;
+use BitBag\SyliusVueStorefront2Plugin\DataGenerator\Generator\Entity\TaxonGenerator;
+use PhpSpec\ObjectBehavior;
+use Prophecy\Argument;
+use Sylius\Component\Core\Model\TaxonInterface;
+use Sylius\Component\Taxonomy\Model\TaxonTranslationInterface;
+
+class TaxonGeneratorSpec extends ObjectBehavior
```
Comment (senghe) [inline spec/DataGenerator/Generator/Entity/TaxonGeneratorSpec.php:23]: All specs should be marked as final.

File: src/DataGenerator/ContextModel/Bulk/BulkContext.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace BitBag\SyliusVueStorefront2Plugin\DataGenerator\ContextModel\Bulk;
+
+use BitBag\SyliusVueStorefront2Plugin\DataGenerator\ContextModel\Entity\EntityContextInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+
+class BulkContext implements BulkContextInterface
```
Comment (senghe) [inline src/DataGenerator/ContextModel/Bulk/BulkContext.php:15]: Could you check whether you've all finals in place?

File: src/DataGenerator/ContextModel/Entity/ProductContext.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    public function getChannel(): ChannelInterface
+    {
+        return $this->channel;
+    }
+
+    public function entityName(): string
+    {
+        return 'Product';
```
Comment (senghe) [inline src/DataGenerator/ContextModel/Entity/ProductContext.php:31]: I would rather use Product::class (as a constant from Product entity class)

File: src/DataGenerator/ConsoleCommand/BulkDataGenerator.php [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        }
+
+        if ($wishlistsQty > 0 && $productsQty > 0) {
+            $productsPerWishlistQty = $this->askForInteger(
+                'Max products per wishlist',
+                self::DEFAULT_PRODUCTS_PER_WISHLIST_QTY
+            );
+        }
+
+        return new DataGeneratorCommandContext(
```
Comment (senghe) [inline src/DataGenerator/ConsoleCommand/BulkDataGenerator.php:143]: I would put this into the context factory class (can be in a new method).

File: src/DataGenerator/ContextModel/Entity/WishlistContextInterface.php [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<?php
+/*
+ * This file was created by developers working at BitBag
+ * Do you need more information about us and what we do? Visit our https://bitbag.io website!
+ * We are hiring developers from all over the world. Join us and start your new, exciting adventure and become part of us: https://bitbag.io/career
+ */
+
+namespace BitBag\SyliusVueStorefront2Plugin\DataGenerator\ContextModel\Entity;
```
Comment (senghe) [inline src/DataGenerator/ContextModel/Entity/WishlistContextInterface.php:8]: Could you check whether all declare statements are in place?

File: src/DataGenerator/Doctrine/Repository/TaxonRepository.php [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+*/
+
+declare(strict_types=1);
+
+namespace BitBag\SyliusVueStorefront2Plugin\DataGenerator\Doctrine\Repository;
+
+use Sylius\Bundle\TaxonomyBundle\Doctrine\ORM\TaxonRepository as BaseTaxonRepository;
+use Sylius\Component\Core\Model\TaxonInterface;
+
+final class TaxonRepository extends BaseTaxonRepository implements TaxonRepositoryInterface
```
Comment (senghe) [inline src/DataGenerator/Doctrine/Repository/TaxonRepository.php:16]: The queries from repository should be tested with PHPUnit integration tests. This applies for all repository classes.

File: src/DataGenerator/Generator/Bulk/AbstractBulkGenerator.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    public function __construct(
+        EntityManagerInterface $entityManager,
+        GeneratorInterface $generator,
+    ) {
+        $this->entityManager = $entityManager;
+        $this->generator = $generator;
+    }
+
+    public function generate(ContextInterface $context): void
```
Comment (senghe) [inline src/DataGenerator/Generator/Bulk/AbstractBulkGenerator.php:34]: I would spec this method, as there is some logic.

File: src/DataGenerator/Resources/services/cli.xml [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+Do you need more information about us and what we do? Visit our https://bitbag.io website!
+We are hiring developers from all over the world. Join us and start your new, exciting adventure and become part of us: https://bitbag.io/career
+-->
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd"
+>
+    <services>
+        <defaults autowire="false" autoconfigure="false" public="true" />
```
Comment (senghe) [inline src/DataGenerator/Resources/services/cli.xml:14]: Does it need to be public?

File: src/Resources/services/factories.xml [hunk 10]
```diff
@@ -8,6 +8,10 @@ We are hiring developers from all over the world. Join us and start your new, ex
 
 <container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+    <imports>
```
Comment (senghe) [inline src/Resources/services/factories.xml:11]: If you're creating separate component, I would create one `services.xml` file inside, importing all dependent xmls, and then import this whole `services.xml` file in global plugin `services.xml` file.

**Review by senghe — CHANGES_REQUESTED** [review]

File: src/DataGenerator/ConsoleCommand/BulkDataGenerator.php (hunk 5 above)
Comment (klloz) [reply src/DataGenerator/ConsoleCommand/BulkDataGenerator.php]: I've created separate factory for `DataGeneratorCommandContext` in order not to violate SRP.

File: src/DataGenerator/Generator/Bulk/AbstractBulkGenerator.php (hunk 8 above)
Comment (klloz) [reply src/DataGenerator/Generator/Bulk/AbstractBulkGenerator.php:34]: It's an abstract class, so I wrote specs for children's implementations of this method.

File: src/DataGenerator/Resources/services/cli.xml (hunk 9 above)
Comment (klloz) [reply src/DataGenerator/Resources/services/cli.xml]: No :)

**Review by senghe — APPROVED** [review]
All things were discussed on-side with @klloz and they've been fixed.

**Pull request merged**